File: 2025-09-16 - LstDupEncadeada/Exemplo_Python/ListaDuplaEncadeada.py
# ------------------------------------------------------------------------
# Classe ListaDuplaEncadeada
# Implementa a interface ListaDuplaEncadeadaInterface
# ------------------------------------------------------------------------

from Node import Node
from ListaDuplaEncadeadaInterface import ListaDuplaEncadeadaInterface

class ListaDuplaEncadeada(ListaDuplaEncadeadaInterface):
   def __init__(self) -> None:
      self.head: Node | None = None
      self.tail: Node | None = None
      self.tamanho: int = 0

   def InserirNoInicio(self, codigo: str, nome: str, carga: int) -> bool:
      try:
         novo = Node(codigo, nome, carga)
         if self.EstaVazia():
            self.head = self.tail = novo
         else:
            novo.no_proximo = self.head
            self.head.no_anterior = novo
            self.head = novo
         self.tamanho += 1
         return True
      except:
         return False

   def InserirNoFim(self, codigo: str, nome: str, carga: int) -> bool:
      try:
         novo = Node(codigo, nome, carga)
         if self.EstaVazia():
            self.head = self.tail = novo
         else:
            self.tail.no_proximo = novo
            novo.no_anterior = self.tail
            self.tail = novo
         self.tamanho += 1
         return True
      except:
         return False

   def InserirNaPosicao(self, codigo: str, nome: str, carga: int, posicao: int) -> bool:
      if posicao < 0 or posicao > self.tamanho:
         return False
      if posicao == 0:
         return self.InserirNoInicio(codigo, nome, carga)
      if posicao == self.tamanho:
         return self.InserirNoFim(codigo, nome, carga)
      try:
         novo = Node(codigo, nome, carga)
         atual = self.head
         for _ in range(posicao - 1):
            atual = atual.no_proximo
         proximo = atual.no_proximo
         novo.no_proximo = proximo
         novo.no_anterior = atual
         atual.no_proximo = novo
         if proximo:
            proximo.no_anterior = novo
         self.tamanho += 1
         return True
      except:
         return False

   def RemoverPorCodigo(self, codigo: str) -> bool:
      if self.EstaVazia():
         return False
      atual = self.head
      while atual and atual.codigo_disciplina != codigo:
         atual = atual.no_proximo
      if not atual:
         return False
      if atual == self.head:
         self.head = atual.no_proximo
         if self.head:
            self.head.no_anterior = None
         else:
            self.tail = None
      elif atual == self.tail:
         self.tail = atual.no_anterior
         if self.tail:
            self.tail.no_proximo = None
         else:
            self.head = None
      else:
         atual.no_anterior.no_proximo = atual.no_proximo
         atual.no_proximo.no_anterior = atual.no_anterior
      self.tamanho -= 1
      return True

   def BuscarPorCodigo(self, codigo: str) -> Node | None:
      atual = self.head
      while atual:
         if atual.codigo_disciplina == codigo:
            return atual
         atual = atual.no_proximo
      return None

   def AtualizarDisciplina(self, codigo: str, novo_nome: str, nova_carga: int) -> bool:
      alvo = self.BuscarPorCodigo(codigo)
      if alvo:
         alvo.nome_disciplina = novo_nome
         alvo.carga_horaria = nova_carga
         return True
      return False

   def ImprimirLista(self) -> None:
      print("---- Conteúdo da Lista ----")
      atual = self.head
      idx = 0
      while atual:
         print(f"[{idx}] Código: {atual.codigo_disciplina} | Nome: {atual.nome_disciplina} | Carga Horária: {atual.carga_horaria}")
         atual = atual.no_proximo
         idx += 1
      print("---------------------------")

   def EstaVazia(self) -> bool:
      return self.tamanho == 0

   def Tamanho(self) -> int:
      return self.tamanho

File: 2025-09-16 - Lista Dup Encadeada/Exemplo_C#/ListaDuplaEncadeada.cs
// ------------------------------------------------------------------------
// Classe ListaDuplaEncadeada
// Implementa a interface ListaDuplaEncadeadaInterface
// ------------------------------------------------------------------------

using System;

namespace Estruturas
{
   public class ListaDuplaEncadeada : ListaDuplaEncadeadaInterface
   {
      private Node head;
      private Node tail;
      private int tamanho;

      public ListaDuplaEncadeada()
      {
         head = null;
         tail = null;
         tamanho = 0;
      }

      public bool InserirNoInicio(string codigo, string nome, int cargaHoraria)
      {
         try
         {
            Node novo = new Node(codigo, nome, cargaHoraria);
            if (EstaVazia())
            {
               head = tail = novo;
            }
            else
            {
               novo.NoProximo = head;
               head.NoAnterior = novo;
               head = novo;
            }
            tamanho++;
            return true;
         }
         catch
         {
            return false;
         }
      }

      public bool InserirNoFim(string codigo, string nome, int cargaHoraria)
      {
         try
         {
            Node novo = new Node(codigo, nome, cargaHoraria);
            if (EstaVazia())
            {
               head = tail = novo;
            }
            else
            {
               tail.NoProximo = novo;
               novo.NoAnterior = tail;
               tail = novo;
            }
            tamanho++;
            return true;
         }
         catch
         {
            return false;
         }
      }

      public bool InserirNaPosicao(string codigo, string nome, int cargaHoraria, int posicao)
      {
         if (posicao < 0 || posicao > tamanho) return false;
         if (posicao == 0) return InserirNoInicio(codigo, nome, cargaHoraria);
         if (posicao == tamanho) return InserirNoFim(codigo, nome, cargaHoraria);

         try
         {
            Node novo = new Node(codigo, nome, cargaHoraria);
            Node atual = head;
            for (int i = 0; i < posicao - 1; i++)
               atual = atual.NoProximo;

            Node proximo = atual.NoProximo;
            novo.NoProximo = proximo;
            novo.NoAnterior = atual;
            atual.NoProximo = novo;
            if (proximo != null)
               proximo.NoAnterior = novo;

            tamanho++;
            return true;
         }
         catch
         {
            return false;
         }
      }

      public bool RemoverPorCodigo(string codigo)
      {
         if (EstaVazia()) return false;

         Node atual = head;
         while (atual != null && atual.CodigoDisciplina != codigo)
            atual = atual.NoProximo;

         if (atual == null) return false;

         if (atual == head)
         {
            head = atual.NoProximo;
            if (head != null) head.NoAnterior = null;
            else tail = null;
         }
         else if (atual == tail)
         {
            tail = atual.NoAnterior;
            if (tail != null) tail.NoProximo = null;
            else head = null;
         }
         else
         {
            atual.NoAnterior.NoProximo = atual.NoProximo;
            atual.NoProximo.NoAnterior = atual.NoAnterior;
         }

         tamanho--;
         return true;
      }

      public Node BuscarPorCodigo(string codigo)
      {
         Node atual = head;
         while (atual != null)
         {
            if (atual.CodigoDisciplina == codigo)
               return atual;
            atual = atual.NoProximo;
         }
         return null;
      }

      public bool AtualizarDisciplina(string codigo, string novoNome, int novaCarga)
      {
         Node alvo = BuscarPorCodigo(codigo);
         if (alvo != null)
         {
            alvo.NomeDisciplina = novoNome;
            alvo.CargaHoraria = novaCarga;
            return true;
         }
         return false;
      }

      public void ImprimirLista()
      {
         Console.WriteLine("---- Conteúdo da Lista ----");
         Node atual = head;
         int idx = 0;
         while (atual != null)
         {
            Console.WriteLine($"[{idx}] Código: {atual.CodigoDisciplina} | Nome: {atual.NomeDisciplina} | Carga Horária: {atual.CargaHoraria}");
            atual = atual.NoProximo;
            idx++;
         }
         Console.WriteLine("---------------------------");
      }

      public bool EstaVazia() => tamanho == 0;

      public int Tamanho() => tamanho;
   }
}
